### ISSUE TITLE
Unity 2019.4.x and 2020.1.x not working without -nographics flag - Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by Alriightyman** [game-ci/unity-actions#84]
I've been trying to get tests to run but came across this error:
```
###########################
#   Testing in EditMode   #
###########################

Segmentation fault (core dumped)
[Licensing::Module] Channel doesn't exist: "LicenseClient-root"
[Licensing::Module] Successfully launched the LicensingClient (PId: 81)
[SignatureVerifier] Application signature verification not supported on this platform.
[Licensing::Module] Successfully connected to LicensingClient on channel: "LicenseClient-root" (connect: 0.81s, validation: 0.07s, handshake: 0.00s)
[Licensing::Module] Connected to LicensingClient (PId: 81, launch time: 0.00, total connection time: 0.88s)
Entitlement-based licensing initiated
[LicensingClient] Licenses updated successfully
Unexpected exit code 139

###########################
#   Testing in PlayMode   #
###########################

Segmentation fault (core dumped)
[Licensing::Module] Channel doesn't exist: "LicenseClient-root"
[Licensing::Module] Successfully launched the LicensingClient (PId: 136)
[SignatureVerifier] Application signature verification not supported on this platform.
[Licensing::Module] Successfully connected to LicensingClient on channel: "LicenseClient-root" (connect: 0.57s, validation: 0.07s, handshake: 0.00s)
[Licensing::Module] Connected to LicensingClient (PId: 136, launch time: 0.00, total connection time: 0.64s)
Entitlement-based licensing initiated
[LicensingClient] Licenses updated successfully
Unexpected exit code 139
```
Seems to be a Seg Fault error. A way to fix this was mentioned here:
https://gitlab.com/gableroux/unity3d-gitlab-ci-example/-/issues/121
This suggests adding the `-nographics` flag to all CI scripts. 

I would assume adding that flag to this file:
`unity-test-runner/action/steps/run_tests.sh`
when executing the tests would fix the issue.

I am unsure as to why you would want to run this *with* graphics, however, I am new to running CI and Unity in this way so maybe there is a reason for this flag to be omitted from those calls?  If so, maybe a variable that can be set and passed?  What are your thoughts?

**Comment by webbertakken** [comment]
Thanks @osum4est for confirming that also 2020 is built without `-nographics` flag in unity cloud build

The arguments look the same between 2019 and 2020 builds.

```
 COMMAND LINE ARGUMENTS:
/UNITY_PATH/Unity/macos/published/Unity-2020_1_0f1/Unity.app/Contents/MacOS/Unity
-batchmode
-skipMissingProjectID
-skipMissingUPID
-projectPath
/BUILD_PATH/project-path
-logFile
-buildconfig
-bvrbuildtarget
win64
-upmServerPort
10010
-buildTarget
standalonewindows64
-quit
``` 

This effectively means that Unity made an undocumented change that caused our docker images to fail.

We will have to try to find out what that change is.

**Comment by webbertakken** [comment]
Possibly also related: https://github.com/webbertakken/unity-builder/issues/136

**Comment by ncipollo** [comment]
Hey @webbertakken,

Not sure if this is helpful but I updated one of my test projects to 2020 and created a docker image which can be used for testing commands within the docker container:

https://github.com/ncipollo/unity-ci-test/

`./docker-build.sh` creates the container, then `docker-run.sh` runs it. I included super simplified versions of your CI scripts. They are included via a volume so you can play around with the scripts without needing to rebuild the container each time.

So far the only thing which seems to prevent the segmentation fault is `-nographics` 😬 . I'll keep playing with it to see if I can get it to output more meaningful error information somehow.

**Comment by ncipollo** [comment]
> Thanks @osum4est for confirming that also 2020 is built without `-nographics` flag in unity cloud build
> 
> The arguments look the same between 2019 and 2020 builds.
> 
> ```
>  COMMAND LINE ARGUMENTS:
> /UNITY_PATH/Unity/macos/published/Unity-2020_1_0f1/Unity.app/Contents/MacOS/Unity
> -batchmode
> -skipMissingProjectID
> -skipMissingUPID
> -projectPath
> /BUILD_PATH/project-path
> -logFile
> -buildconfig
> -bvrbuildtarget
> win64
> -upmServerPort
> 10010
> -buildTarget
> standalonewindows64
> -quit
> ```
> 
> This effectively means that Unity made an undocumented change that caused our docker images to fail.
> 
> We will have to try to find out what that change is.

Just noticed something interesting in this snippet, it looks like this is running on a `macOS` machine. The containers we are using run on 'Ubuntu 18.04.4 LTS'. This may be that command functions correctly without `-nographic` (i.e this might be a linux specific issue in unity)

**Comment by ncipollo** [comment]
This seems to be a viable workaround, at least in the test runner action:

```yml
build:
  runs-on: ubuntu-latest
  steps:
  - uses: actions/checkout@v1
  - name: 'run unity tests'
    uses: webbertakken/unity-test-runner@v1.6
    env:
      UNITY_LICENSE: ${{ secrets.UNITY_LICENSE }}
    with:
      customParameters: '-nographics' # <--- This is the important bit
      unityVersion: 2020.1.2f1
```

**Comment by webbertakken** [comment]
> This seems to be a viable workaround, at least in the test runner action:

Respectfully, this is not a good fix as discussed above.

**Comment by ncipollo** [comment]
@webbertakken ,
To clarify - I am not proposing this as a solution. This workaround worked for me in my project and I thought I would share it for folks who are also using 2020.1.X. While `-nographics` may not work for everyone, at least having a documented workaround for CI projects which don't require hardware acceleration could be useful.

**Comment by webbertakken** [comment]
You are right, and technically I should work for builder too. I am just afraid for the bug reports that happen because people don't remove the flag again and bug reports caused by this flag. 

It's easy to look for shortcuts but it kind of stops people from thinking along, which means it's gonna cost me even more of my free time to fix it properly.

**Comment by Alriightyman** [comment]
Very true. Unfortunately, it seems like it is very much an issue with the 2020 version of Unity not just the released versions but the beta versions as well ( you can look through the gitlab issues through the initial post above). But @webbertakken is correct. I'm sure there will be bug reports about this in the future because someone forgot they added this flag when they bumped into this issue. To me, it sounds like this bug should actually be reported to Unity although, I am not sure where to do that at the moment. 
As a side note, I should probably remove my forked version before more people use that thinking it's a working version when it's really not. Thoughts?

**Comment by webbertakken** [comment]
> As a side note, I should probably remove my forked version before more people use that thinking it's a working version when it's really not. Thoughts?

Up to you. At least when people switch from your fork back to the original they'll not have to revert any configuration.

**Comment by ncipollo** [comment]
>  I'm thinking maybe we'll have to up the viewport size inside the docker container or something similar.
If you're talking about the resolution which is passed into `xvfb-run`, I tried a bunch of different options and sadly they all still led to a segmentation fault :( . If you have had some specific resolutions you wanted to test I can fire up my test project and try them out. 

It might also be worth filing a bug to unity. This seems like an issue which is specific to their 2020.x linux builds. If no one has done that yet I could give it a shot. As I mentioned in an earlier comment I have docker image + scripts which make this very easy to reproduce.

**Comment by pandariusArt** [comment]
Hi guys, i'm new here so please excuse me in advance if my intervention can be seen as pointless.

Just to say that i can also reproduce the docker error code 139 with the action file provided here : https://unity-ci.com/docs/github/getting-started.

And it happens on the build stage with a fresh project on 2020.1.

**Comment by webbertakken** [comment]
@pandariusArt you've found the correct issue for it. We're discussing exactly that 🙂

**Comment by pandariusArt** [comment]
@webbertakken thank you for your answer, how can i contribute and help ?

**Comment by webbertakken** [comment]
@pandariusArt I really appreciate your intention to contribute. Note that it might not be the easiest problem to solve per se. The issue describes most aspects of the problem at least on a high level. When you've read all that here's what you could do to contribute.

If you have knowledge about docker and linux, you might be able to help getting more information out of the error that we're getting. Ultimately we could really use a Windows docker container and maybe even MacOS docker container that can run unity, to rule out that the issue is related to linux (perhpas it's related to unity in docker in general). See [gableroux/unity3d](https://gitlab.com/gableroux/unity3d/) which contains the docker image that is used by this project.

If you're good at explaining things it might be helpful to create that bug that @ncipollo suggested we create.

If you're good at fiddling around you might be able to find other solutions than `-nographics` flag to run the container.

And we're generally always looking for fresh perspectives.

Cheers!

**Comment by pandariusArt** [comment]
> @pandariusArt I really appreciate your intention to contribute. Note that it might not be the easiest problem to solve per se. The issue describes most aspects of the problem at least on a high level. When you've read all that here's what you could do to contribute.
> 
> * If you have knowledge about docker and linux, you might be able to help getting more information out of the error that we're getting.
> * Ultimately we could really use a Windows docker container and maybe even MacOS docker container that can run unity, to rule out that the issue is related to linux (perhpas it's related to unity in docker in general).
> * Any contributions to [gableroux/unity3d](https://gitlab.com/gableroux/unity3d/) which contains the docker image, also help this project directly.
> * If you're good at explaining things it might be helpful to create that bug that @ncipollo suggested we create.
> * If you're good at fiddling around you might be able to find other solutions than `-nographics` flag to run the container.
> 
> And we're generally always looking for fresh perspectives.
> 
> Cheers!

@webbertakken thank you for this detailled answer. I'll check if i can do something with my humble knowledge on Docker with Linux. I'll keep you guys updated if i find anything new and helpful.

Have a nice day ;)

**Comment by ncipollo** [comment]
Filed a bug with unity last night - https://fogbugz.unity3d.com/default.asp?1272554_hjtc0dn7riqjk9ei

**Comment by Kleptine** [comment]
Thanks for the pointer! Man, it's very hard to find these threads on Google.

From our point of view, -nographics is a perfectly fine fix. All the tests I've run so far work fine under no-graphics. But I definitely appreciate the goal of raising this up to Unity as a bug. The Editor should never segfault.

That said, xvfb was always a really weird hack, in my opinion. So for me it's just trading one hack for another. 🤷

**Comment by ncipollo** [comment]
The unity team responded to my bug: 

```
Hi,

Thank you for contacting Unity about your issue.

Currently, we do not officially support running Unity inside a Docker container. But the issue you have reported is known and may or may not get fixed due to Docker not being officially supported.

You can read more about Docker support here: https://forum.unity.com/threads/command-line-manuallicensefile-on-linux.889129/#post-5909540

If you have further questions, feel free to contact us.

Best regards,
Augustinas
Customer QA Team 
```

tldr; It sounds like they don't officially support running in a container, though it sounds like they are aware of the crash (so maybe they will fix it 😬 ).

**Comment by thorgeirk11** [comment]
Is there any news on this issue?

We are experiencing issues with both 2020.1.2 and 2019.4.9f1 both end up failing without any error. 
See our original post #83 for more details. We have tried adding the '-nographics' flag, but it had no effect.

**Comment by webbertakken** [comment]
> Is there any news on this issue?
> 
> We are experiencing issues with both 2020.1.2 and 2019.4.9f1 both end up producing no build without any error.
> See our original post #83 for more details. We have tried adding the '-nographics' flag, but it had no effect.

Sorry to hear that @thorgeirk11,

Unfortunately we do not have a proper solution yet as we can't simply compile unity itself, but if you have any suggested approaches we'd be more than happy to consider them.

Right now `-nographics` seems to be peoples go-to workaround. You're actually the first person reporting that's not working. Are you sure you are getting the same errors? Feel free to create a separate issue for this and give as much detail as possible about the circumstances under which even `-nographics` does not work.

**Comment by louis030195** [comment]
I can't make build works too on 2020.1.4f1 despite the "nographics" (seg fault), it solved the unit test though. see my [config](https://github.com/louis030195/niwrad/blob/master/.github/workflows/main.yml#L87).
Anyway I've been using unity on Linux for years, I can barely compile anything locally even with UI (only Linux build works), I have segfault for everything Unity-cli, any version.

Maybe a Linux image using wine with Unity is an option ? Otherwise I guess it would be something with Windows/Mac I guess ...

**Comment by webbertakken** [comment]
The segfault is immediately after setting graphics device to null and while loading your assets. I'm not certain but it looks like some of your assets require graphics card rendering (which is what you disable with `-nographics` as far as I know.

Feel free to make a separate issue for this if you want to discuss it. Remember to provide as much information as possible about how your project differs from an empty project.

**Comment by nicholas-maltbie** [comment]
Hey, I'm working with this great tool and just wanted to note that I ran into this issue. Adding the flag `-nographics` to my github action as part of the test workflow. Made it work perfectly. No new issues just wanted to say thanks for the great documentation adn post an example if it helps for future development. I was using 2019 before but decided to upgrade my project to use 2020.1.5f1 Here are the logs of the two github actions if you needed an example.

Previous log without flag: [log](https://pipelines.actions.githubusercontent.com/Zztm2TlUCzjEhWEpQBbxj2xOo9xG30sAxQ8hJ291QeMFqz2JMt/_apis/pipelines/1/runs/273/signedlogcontent/3?urlExpires=2020-09-14T04%3A28%3A41.4793268Z&urlSigningMethod=HMACV1&urlSignature=qdv4wQ1%2FDnN6rCDmimJQ2teIOjryP3Oa7W2YY7kwDRU%3D)
Previous workflow without flag: [workflow](https://github.com/nicholas-maltbie/PropHunt/actions/runs/253011986/workflow#L52)

Log with `-nographics` flag: [log](https://pipelines.actions.githubusercontent.com/Zztm2TlUCzjEhWEpQBbxj2xOo9xG30sAxQ8hJ291QeMFqz2JMt/_apis/pipelines/1/runs/276/signedlogcontent/3?urlExpires=2020-09-14T04%3A28%3A10.3592403Z&urlSigningMethod=HMACV1&urlSignature=tFfFOf1xO7gRB5SQB8zjoVtA9dz0h%2BUp69%2BB3oFmHkA%3D)
Workflow with `-nographics` flag: [workflow](https://github.com/nicholas-maltbie/PropHunt/actions/runs/253176106/workflow#L52)

I'm going to try and add some playmode tests to my project whenever I figure out how to do that, I don't think graphical tests are necessary (I was able to run them locally from powershell CLI so I would assume it would work with ubuntu, just making a fun project to learn more about unity and ECS now).

**Comment by webbertakken** [comment]
Ok lets vote: Yay or nay

We could add a parameter to the action for `graphics`, which defaults to `false`, as it seems to be the only solution we could come up with for now. Setting it to true will remove the `-nographics` flag.

What do you think?

**Comment by Alriightyman** [comment]
As we've discussed, this only solves part of the problem; some tests will not work properly such as GI.  If we DO add a flag, I'd hope you'd make the effect it has clear in the ReadMe or any documentation related to this. 

So, I vote yes as long as it clearly states the limitations of this flag with certain Unity builds (version 2020 and possibly beyond).

Also, should it default to true since that would be the original behavior?  There are most likely people still working in Version 2019 which behaves correctly.  Or are you thinking ahead assuming Unity won't bother repairing the potential issue on their end (also assuming it's something on their end) or that "we" cannot find a fixable solution. 

Thoughts?

**Comment by webbertakken** [comment]
This has been my standpoint from the start, but I'm only seeing other people being perfectly satisfied with the -nographics flag.

If we add this option, it should actually solve a problem, if we set it to default to true then we might as well just stick to the current solution. I'll count that as a vote for no. Thanks. I really appreciate your input!

**Comment by Alriightyman** [comment]
I'm ok with hard coding it. But just make sure to clearly add information about it within the readme. This way, if you close this ticket, someone doesn't open another ticket related to this issue sometime in the future.  Just my thoughts.

**Comment by crassSandwich** [comment]
I've also been running into this issue, but adding the '-nographics' flag didn't fix it for me. I haven't had time to do a full investigation, but it seems to be linked to license activation, as per this discussion on the Unity forums:

https://forum.unity.com/threads/cant-get-license-in-headless-nographics-mode-timeout-error.479960/

Seems like Unity spawns a window during the activation process, and there's a workaround for that in the post. For now I'm going to stick with 2019.3, but this looks like something to keep in mind if the decision is to go with adding -nographics

**Comment by webbertakken** [comment]
This is partially solved in [v2 alpha](https://github.com/webbertakken/unity-builder/releases) from builder but if builds still fail (build always provides `-nographics`) it's related to a bug in Unity itself, which no longer works without `-nographics` from version 2019.4 on.

**Comment by Alriightyman** [comment]
@webbertakken What exactly does “partially solved” mean here?

**Comment by webbertakken** [comment]
I describe what it means in the same message as the text you're quoting. I'm not exactly sure which part is unclear. Let me rephrase;

- It's a bug in Unity, that doesn't run on Linux machines in batch mode in any version from 2019.4 on.
- There's a not-so-nice workaround of always passing `-nographics`, which will only work if your build doesn't require GPU processing during build time.

**Comment by Alriightyman** [comment]
Sorry, I was referring to the changes in v2 alpha. All I see is removing duplicated flags; I don’t see how those are related. Unless the change in docker images is why you are referring to.

**Comment by webbertakken** [comment]
The change in docker images is what v2 alpha is all about. Please refer to the release notes of the first alpha release and all release notes of the docker repo for all the details.

**Comment by RDeluxe** [comment]
Just chiming in to say that our build works with 2020.1 but only WITHOUT the `-nographics` flag.
With the flag, we got a segfault during the build phase

**Comment by webbertakken** [comment]
Please when reporting specific builds to work without the flag, be sure to mention which version of the action and which exact version of Unity editor you're using.

**Comment by RDeluxe** [comment]
Unity version : 2020.1.16f1

I used this custom Dockerfile : 

```
FROM unityci/editor:ubuntu-2020.1.16f1-android-0.7.0
RUN apt-get install -y vim git git-lfs htop 
RUN git lfs install
RUN mkdir /root/.ssh 
COPY id_rsa /root/.ssh/
RUN chmod 600 /root/.ssh/id_rsa
```

Cloned my repo, and launched the build like this :


```
unity-editor \
  -logfile /dev/stdout \
  -quit \
  -customBuildName "Builds" \
  -projectPath "./cityinvaders" \
  -buildTarget "android" \
  -customBuildTarget "android" \
  -executeMethod "LoneStoneStudio.CityInvaders.CloudBuildMethods.CloudBuild" \
  -androidKeystorePass <...> \
  -androidKeyaliasName <...> \
  -androidKeyaliasPass <...> \
  -env STAGING  \
  -buildPath builds 
```

**Comment by RDeluxe** [comment]
Hey there !

Just chiming in to say that my problem may not be linked to the nographics flag, or in a non obvious way.

Our game was not compiling on Unity Cloud build either, however Unity Cloud logs were more verbose (because it does not launch a C# script in the editor, but instead a command line).
It seems our project was not compiling when targeting Android AMR64 because the project size was too big : https://forum.unity.com/threads/android-builds-failing-when-script-debugging-is-enabled.1027357/

Here is our thread on Unity forums : https://forum.unity.com/threads/il2cpp-build-failure.1055486/

The project compiles fine on windows, and disabling ARM64 target (keeping only ARMv7) makes the build work with unity-builder.

If I disable Debugging, I can re-enable ARM64 as a target and Unity Cloud build will work (\o/), however Unity-builder will still fails. I honestly have no idea why, and I'm too tired to look further. If we one day upgrade to 2021.1 (which seems to embark a fix) I'll try unity-builder again on this project.

Thanks 🙏

**Comment by Warhammer4000** [comment]
Here are a few observations from me to help on the context

1) One of my project was being developed in Legacy RP(Default) Unity Version 2019.4(LTS). I used the CI to produce builds for StandaloneWindows64. I believe I was using UnityBuilder@v2-alpha-2. When the Alpha-6 came out I updated the builder it was working for a few days and suddenly it stopped working. I believe that's when the WebGL Issues were fixed regarding some python library. The Build errors were simply "Segmentation Fault, Core Dumped". And few other projects stopped working. I don't remember when but I had the project shift to URP. And It did work for a while and then it didn't. But I'm positive that it stopped working after that WebGL Builder update.

2) Just 2 days ago I tested out a build that removed -nographic flag. And without any configuration change my previous projects started to work again. More Specifically the projects that had URP on it. Because Legacy RP projects never failed so far. Just Removing -nographic flag fixed the segmentation fault problem from 2 of my Projects Target Platform for both was StandaloneWIndows64. They both has URP. 

3) I'm Building the same project for Android and StandaloneOSX without nographic flag. I'm not seeing any problem. When the builder wasn't working I tested for different platforms and it still gave the core dumped seg fault issue. So I have doubt if this is Something Regarding Platform.

4) I used LWRP in one project and didn't update it. It works with -nographic flag

**Comment by webbertakken** [comment]
If indeed it is related to specific packages, or even render pipeline (package?) specifically, that might also mean that https://github.com/game-ci/unity-test-runner/pull/109 should prevent this from happening during license activation step in test runner (not for the test run itself). 

It's also something we have to implement in unity-builder still.

Perhaps we should run some tests:
- for all unity versions
- with `URP` and `HDRP` and `legacy RP` (and `LWRP`?) projects
- with `-nographics` and without `-nographics`

Perhaps we'll start to see the pattern then. 
@mob-sakai is this something you could manage on your mass-test repo?
